src/ic-cdk/src/api/stable.rs

```diff
@@ -161,6 +162,142 @@ impl io::Write for StableWriter {
+            // There is enough room in the buffer to store the new bytes.
+            Ordering::Greater => {
+                self.buffer.extend_from_slice(buf);
+            }
+            // There is enough room in the buffer to store the new bytes, but now it is full so we
+            // need to flush it to stable memory.
+            Ordering::Equal => {
+                self.buffer.extend_from_slice(buf);
+                self.flush()?;
+            }
+            // The new bytes will not fit in the buffer.
+            Ordering::Less => {
+                // We can reduce the calls to grow stable memory by growing to the total known
+                // length here rather than leaving it up to `flush` which will only grow by up to
+                // the length of the buffer.
+                let total_capacity_required =
+                    self.offset + self.buffer.len() as u64 + buf.len() as u64;
+                self.grow_to_capacity_bytes(total_capacity_required)?;
+
+                // If the new bytes exceed the capacity remaining + the starting capacity then we
+                // must flush everything now since we will not be able to fit the bytes into the
+                // buffer.
+                // Otherwise we fill the buffer, flush it, then populate the buffer again with the
+                // remaining bytes.
+                if buf.len() > self.buffer.capacity() + buffer_capacity_remaining {
+                    self.flush()?;
+                    self.memory.stable64_write(self.offset, buf);
+                    self.offset += buf.len() as u64;
+                } else {
+                    self.buffer
+                        .extend_from_slice(&buf[..buffer_capacity_remaining]);
+                    let remaining_to_write = &buf[buffer_capacity_remaining..];
+                    self.flush()?;
+                    self.buffer.extend_from_slice(remaining_to_write);
+                }
+            }
+        }
+
+        Ok(buf.len())
+    }
+
+    /// Attempts to grow the memory by adding new pages.
+    pub fn grow(&mut self, added_pages: u64) -> Result<(), StableMemoryError> {
+        let old_page_count = self.memory.stable64_grow(added_pages)?;
+        self.capacity = old_page_count + added_pages;
+        Ok(())
+    }
+
+    pub fn flush(&mut self) -> Result<(), StableMemoryError> {
+        if !self.buffer.is_empty() {
+            self.grow_to_capacity_bytes(self.offset + self.buffer.len() as u64)?;
+            self.memory.stable64_write(self.offset, &self.buffer);
+            self.offset += self.buffer.len() as u64;
+            self.buffer.clear();
+        }
+
+        Ok(())
+    }
+
+    fn grow_to_capacity_bytes(
+        &mut self,
+        required_capacity_bytes: u64,
+    ) -> Result<(), StableMemoryError> {
+        let required_capacity_pages =
+            (required_capacity_bytes + WASM_PAGE_SIZE_IN_BYTES - 1) / WASM_PAGE_SIZE_IN_BYTES;
+        let current_pages = self.capacity as u64;
```

Comment:
Would it make sense to check that `self.capacity` matches the current stable memory size? To protect against multiple users of the stable memory.

Reply:
We want to minimise calls to `stable_size`, hence the need for the `capacity` field, but each time we grow stable memory we are told the latest size and set `capacity` to that new value.
If there are multiple users of stable memory then `capacity` could be incorrectly too low, but this won't cause any issues, it will just cause `stable_grow` to be called prematurely.

Reply:
Wouldn't that cause data corruption since we would be overwriting the bytes written by another user?

Reply:
Each canister has its own stable memory so it is up to the canister's devs to ensure that if they are using multiple writers within their canister then they don't overwrite each other's data.
Also, canisters are single threaded so I don't think there's really a use case for ever having multiple writers (maybe if you want to write at different offsets and want to avoid the call to `stable_size` each time you switch offset, but it seems unlikely anyone would do that).

The downside is that if we were to add a call to `stable_size` within `write` then each write would be 2 system calls (or more if we need to grow) rather than 1.
What would we even do in the case where the `capacity` field doesn't match `stable_size`? That would only tell us that the canister has separately called `stable_grow`, but doesn't tell us if we'd be overwriting data that we shouldn't be.

Reply:
In fact whenever writing to stable memory devs will need to ensure they don't overwrite data which they want to keep, regardless of if the writes are happening 'concurrently' or not.

Reply:
Yep, I meant this as a help for devs to ensure that they are not mixing multiple writer/serialization libraries.